unicon_backend/evaluator/tasks/programming.py

```diff
@@ -23,23 +33,40 @@ class File(BaseModel):
 
 
 class StepType(str, Enum):
-    PY_RUN_FUNCTION = "PY_RUN_FUNCTION"
+    PY_RUN_FUNCTION = "PY_RUN_FUNCTION_STEP"
 
 
-class Step(BaseModel):
+class Step(CustomBaseModel, abc.ABC, Generic[InputType, OutputType], polymorphic=True):
+    id: int
     type: StepType
 
 
-class PyRunFunctionStep(Step):
+class PyRunFunctionStep(Step[Any, Any]):
     function_name: str
-    arguments: list[str]
+    arguments: list[str | int]
     keyword_arguments: dict[str, str]
 
+    def get_code(self):
+        pass_to_function = self.arguments + \
+            [f"{key}={value}" for key, value in self.keyword_arguments.items()]
+        return f"{self.function_name}({', '.join(map(str, pass_to_function))})"
+
 
 class Testcase(BaseModel):
     id: int
     steps: list[Step]
 
+    def run(self, prefix: str, answer: "Answer"):
+        folder_path = os.path.join(OUTPUT_FOLDER, prefix, str(self.id))
+        os.mkdir(folder_path)
+        artifacts = answer.artifacts
+        run_funcs = [step for step in self.steps if isinstance(
+            step, PyRunFunctionStep)]
+        file = template.render(
+            prepend="", artifacts=artifacts, run_funcs=run_funcs)
+        with open(os.path.join(folder_path, "run.py"), "w") as f:
+            f.write(file)
```

Comment:
thoughts on whether it is better to just skip the step of the serialising it into files and directly send a `multipart/form-data` or `json` payload to our runner?

the `json` payload could look something like:

```json
{
    "files": [ {"file_name": "file_x", "content": ""} ],
    "environment": {},
    "entrypoint": "file_x.py" 
}
```

Reply:
yea this looks good! do you foresee `output` being just the program output or a more consistent response `PASS` or `FAIL` since it will be a testcase

_not sure if which is better but i guess we can keep the runner as simple as possible by containing the logic of resolving test cases in the backend and have the runner just run/execute programs_